packages/runtime/src/widgets/PopupMenu.tsx

```diff
@@ -38,99 +51,145 @@ interface PopupMenuStyles {
-              : null}
-          </React.Fragment>
-        );
+    if (isObject(itemTemplate) && !isEmpty(evaluatedNamedData.namedData)) {
+      const tempItems = evaluatedNamedData.namedData.map((item, index) => {
+        const itm: ItemType = {
+          key: `popupmenu_itemTemplate_${index}`,
+          label: (
+            <CustomScopeProvider value={item as CustomScope}>
+              {EnsembleRuntime.render([itemTemplate.template])}
+            </CustomScopeProvider>
+          ),
+        };
+        return itm;
       });
 
       popupItems.push(...tempItems);
     }
 
-    return (
-      <Menu className={values?.styles?.names} style={{ ...values?.styles }}>
-        {popupItems}
-      </Menu>
-    );
-  }, [values?.items, action, namedData, itemTemplate]);
+    if (values?.showDivider) {
+      for (let i = 1; i < popupItems.length; i += 2) {
+        popupItems.splice(i, 0, { type: "divider" });
+      }
+    }
+
+    return popupItems;
+  }, [
+    values?.items,
+    values?.showDivider,
+    itemTemplate,
+    evaluatedNamedData.namedData,
+    getMenuItem,
+  ]);
 
-  useEffect((): void => {
-    if (values && !values.widget) {
-      return;
+  const getWidget = useCallback(() => {
+    if (!values?.widget) {
+      throw Error("PopupMenu requires a widget to render the anchor.");
     }
-    // clone value so we're not updating the yaml doc
-    const widget = cloneDeep(values?.widget);
-    const actualWidget = unwrapWidget(widget!);
-    setWidgetProps(actualWidget);
-    // Only run once
-  }, []);
+    const widget = cloneDeep(values.widget);
+    const actualWidget = unwrapWidget(widget);
+    return EnsembleRuntime.render([actualWidget]);
+  }, [values?.widget]);
+
+  const handleMenuItemClick = useCallback<NonNullable<MenuProps["onClick"]>>(
+    ({ key }) => {
+      let item;
+      if (key.includes("itemTemplate")) {
+        const itemIndex = toNumber(key.split("_").at(-1));
+        item = evaluatedNamedData.namedData[itemIndex];
+      } else {
+        const itemIndices = key.split("_").filter((_, i) => i > 1);
+        if (itemIndices.length > 1) {
+          for (let i = 1; i < itemIndices.length; i += 2) {
+            itemIndices.splice(i, 0, "items");
+          }
+        }
+        item = get(values?.items, itemIndices) as unknown;
+      }
+
+      action?.callback({ value: item });
+    },
+    [action, evaluatedNamedData.namedData, values?.items],
+  );
 
   return (
-    <>
-      <AntdDropdown overlay={popupMenuItems} trigger={["click"]}>
-        <div>{widgetProps ? EnsembleRuntime.render([widgetProps]) : null}</div>
+    <div ref={rootRef}>
+      <AntdDropdown
+        disabled={values?.enabled === false}
+        menu={{
+          id,
+          items: popupMenuItems,
+          onClick: handleMenuItemClick,
+          style: { overflow: "auto", ...values?.styles },
+        }}
+        trigger={[values?.trigger || DEFAULT_POPUPMENU_TRIGGER]}
+      >
+        <div>{getWidget()}</div>
```

Comment:
What is this extra div element for?

Reply:
can use fragment like other places.
~ was using it as our runtime render method returns an array of ReactNode